WS17_18/05_lifecycle_segues/your assignment/hahn/Yatodoa/Yatodoa/Controller/TagTableViewController.swift

```diff
@@ -0,0 +1,127 @@
+    case created
+}
+
+extension TagType {
+    
+    init(tag: Tag) {
+        self = .created
+    }
+    
+    var headerTitle: String {
+        switch self {
+            case .created: return "Tags"
+        }
+    }
+    
+    static func < (lhs: TagType, rhs: TagType) -> Bool {
+        return lhs.hashValue < rhs.hashValue
+    }
+}
+
+protocol TagTableViewControllerDelegate {
+    func ttvc(_ vc: TagTableViewController, selectedTag tag: Tag?)
+}
+
+class TagTableViewController: UITableViewController {
+    
+    var tags: [Tag] = [] {
+        didSet {
+            print(tags)
+            
+            tagsSections = tags.groupBy(key: { tag in
+                return TagType(tag: tag)
+            }).sorted { (l, r) -> Bool in
+                return l.key < r.key
+            }
+        }
+    }
+    
+    var tagsSections: [(key: TagType, value: [Tag])] = []
+    
+    var delegate: TagTableViewControllerDelegate?
+    
+    var selectedTag: String?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.tags.append(contentsOf: Tags)
+    }
+    
+    private func populate(_ n: Int, tagNames: [String]) {
+        (0..<n).forEach { _ in // populate data
+            let randomIndex = Int(arc4random_uniform(UInt32(tagNames.count)))
+            tags.append(Tag(label: tagNames[randomIndex]))
+        }
+    }
+    
+    // MARK: - Table view data source
+    
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return tagsSections.count
+    }
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return tagsSections[section].value.count
+    }
+    
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath)
+        cell?.accessoryType = UITableViewCellAccessoryType.checkmark
+        selectedTag = cell?.textLabel?.text
+        
+        tableView.reloadData()
+    }
+    
+    override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
+        let cell = tableView.cellForRow(at: indexPath)
+        cell?.accessoryType = UITableViewCellAccessoryType.none
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        // delegate aufrufen
+        let tag = tags.filter { (tag) -> Bool in
+            return tag.label == selectedTag
+        }.first
+        
+        delegate?.ttvc(self, selectedTag: tag)
+        print("viewWillDisappear")
+    }
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let tag = tags[indexPath.row]
+        
+        let cell = UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: "cell")
+        cell.textLabel?.text = tag.label
+        cell.textLabel?.textColor = tag.color
+        
+        if let selectedTag = selectedTag, tag.label == selectedTag {
```

Comment:
diese Zeile ist superwichtig: du hast dich dafür entschieden, bei jedem `didSelect` den accessory type zu setzen und bei jedem `deselect` diesen zu entfernen. das ist gut und funktioniert auch, allerdings nicht beim ersten durchlauf, wenn der vorherigen VC dir den `selectedTag` übergibt. um das checkmark initial zu setzen gibt es eigentlich nur eine Möglichkeit: in `cellForRowAtIndexpath`. 

warum? weil `cellForRowAtIndexPath` die **einzige** Stelle ist, wo du das aussehen bzw. die properties deiner cell setzt. ein property ist der `accessorytype`, den du davon abhängig setzt, ob die aktuelle cell jenen wert entspricht, welcher im `selectedTag` steht.

warum ist das wichtig? weil du jetzt zwei stellen hast, an denen du den accessorytype setzt. das ist nicht schlimm, allerdings hast du mehr Verwaltungsaufwand, weshalb du in `didSelect` auch ein `reloadData` machen musst. das ist seltsam, kommt aber daher, weil du eine Verantwortlichkeit an zwei stellen hast

du hast alles richtig gemacht, hättest aber nachdem du diesen bug entdeckt hast, dass der initiale tag nicht wieder verschwindet, ein refaktoring machen müssen und dich auf "eine zentrale stelle für das setzen des `accessoryTypes`" einigen müssen